feat(kfileupload): new file upload kongponent (#829)
Kong/kongponents: docs/components/file-upload.md

# File Upload

> **KFileUpload** provides a wrapper around general `input` with `type=field` for `file upload` or `image upload`.

Reviewer: ```suggestion
**KFileUpload** provides a wrapper around general `input` with `type='file'` for `file upload` or `image upload`.
```

Typo?

> **KFileUpload** provides a wrapper around general `input` with `type=field` for `file upload` or `image upload`.

Reviewer 2: issue: typos
```suggestion
**KFileUpload** provides a wrapper around an `input` element with `type=file` for `file upload` or `image upload`.
```

 
<KCard>
  <template v-slot:body>

>     <KFileUpload label="Upload File" :label-attributes="{ help: 'I use the KLabel `help` prop' }" help="Additional text." icon="close" :accept="['.yml', '.yaml', '.json', '.md', '.markdown', 'image/gif', 'image/jpg', 'image/jpeg', 'image/ico', 'image/png']" appearance="file" removable />

Reviewer: suggestion: Can we simplify the initial example a bit? I'd recommend stripping it down to just use the `label` and `removable` props.

  </template>
</KCard>

```html

> <KFileUpload label="Upload File" :label-attributes="{ help: 'I use the KLabel `help` prop' }" help="Additional text." icon="close" :accept="['.yml', '.yaml', '.json', '.md', '.markdown', 'image/gif', 'image/jpg', 'image/jpeg', 'image/ico', 'image/png']" appearance="file" removable />

Reviewer 2: I'm not sure I understand what `icon="close"` is doing here?

Author: typo, updated it.

```

## Props

> ### appearance

Reviewer 2: Instead of appearance, I think it makes more sense for this prop to be named `type` since it's actually determining the type of upload, `file` or `image`

> KFileUpload has default appearance of type `file`, and can be switched either to - `file`, `image`.

Reviewer: ```suggestion
KFileUpload has has two supported appearances: `file` (default) and `image`.
```

<KCard>
  <template v-slot:body>

>     <KFileUpload label="Upload File" appearance="file" :label-attributes="{ help: 'I use the KLabel `help` prop' }" help="Additional files can be uploaded from HomePage." icon="close" :accept="['.yml', '.yaml', '.json', '.md', '.markdown', 'image/gif', 'image/jpg', 'image/jpeg', 'image/ico', 'image/png']" removable @file-added="(i) => { fileName = i[0].name, fileSize = i[0].size }" @file-removed="() => { fileName = '', fileSize = '' }" />

Reviewer: ```suggestion
    <KFileUpload label="Upload File" appearance="file" :accept="['.yml', '.yaml', '.json', '.md', '.markdown'" removable @file-added="(i) => { fileName = i[0].name, fileSize = i[0].size }" @file-removed="() => { fileName = '', fileSize = '' }" />
```

Can we simplify a bit?

  </template>
</KCard>

> <div class="mt-6">Emitted value: <pre v-if="fileName && fileSize" class="emitted-value">{{ `File Name: ${fileName}` }} and {{ `File Size: ${fileSize}` }}</pre></div>

Reviewer: I think we don't need this little bit since your example in the `Events` section covers it

```html

> <KFileUpload label="Upload File" appearance="file" :label-attributes="{ help: 'I use the KLabel `help` prop' }" help="Additional files can be uploaded from HomePage." icon="close" :accept="['.yml', '.yaml', '.json', '.md', '.markdown', 'image/gif', 'image/jpg', 'image/jpeg', 'image/ico', 'image/png']" removable />

Reviewer: For readability can we break the code snippets up onto multiple lines? You may need to wrap it in `<div>` so the markdown doesn't break

```

<KCard class="mt-6">
  <template v-slot:body>

>     <KFileUpload appearance="image" icon="image" :accept="['.yml', '.yaml', '.json', '.md', '.markdown', 'image/gif', 'image/jpg', 'image/jpeg', 'image/ico', 'image/png']" placeholder="Upload new image (Max 4 MB)" @file-added="(i) => { imageName = i[0].name, imageSize = i[0].size }" >

Reviewer: ```suggestion
    <KFileUpload appearance="image" icon="image" :accept="['image/gif', 'image/jpg', 'image/jpeg', 'image/ico', 'image/png']" placeholder="Upload new image (Max 4 MB)" @file-added="(i) => { imageName = i[0].name, imageSize = i[0].size }" >
```

Can we simplify a bit?

    </KFileUpload>
  </template>
</KCard>

> <div class="mt-6">Emitted value: <pre v-if="imageName && imageSize" class="emitted-value">{{ `File Name: ${imageName}` }} and {{ `File Size: ${imageSize}` }}</pre></div>

Reviewer: I think we don't need this little bit since it's covered in `Events`

```html
<KFileUpload appearance="image" icon="image" :accept="['.yml', '.yaml', '.json', '.md', '.markdown', 'image/gif', 'image/jpg', 'image/jpeg', 'image/ico', 'image/png']" placeholder="Upload new image (Max 4 MB)" />
```

### label

> String to be used as the `KFileUpload` label when `appearance` is of type `file`.

Reviewer: I haven't looked at the code but I wouldn't restrict this prop to only be allowed with `file` appearance. In the dev portal use case they can choose to not use the prop, but we should allow use with `appearance` `image`

### labelAttributes

> Use the `labelAttributes` prop to configure the **KLabel's** [props](/components/label.html) if using the `label` prop when `appearance` is of type `file`.

Reviewer: Likewise, we shouldn't restrict this props usage based on appearance

<KCard>
  <template v-slot:body>

>     <KFileUpload label="Upload File" :label-attributes="{ help: 'Hello, here is the example' }" icon="close" :accept="['.yml', '.yaml', '.json', '.md', '.markdown', 'image/gif', 'image/jpg', 'image/jpeg', 'image/ico', 'image/png']" appearance="file" removable />

Reviewer: ```suggestion
    <KFileUpload label="Upload File" :label-attributes="{ help: 'Hello, here is the example' }" :accept="['.yml', '.yaml', '.json', '.md', '.markdown']" appearance="file" removable />
```

Can we simplify?

  </template>
</KCard>

```html
<KFileUpload label="Upload File" :label-attributes="{ help: 'Hello, here is the example' }" icon="close" :accept="['.yml', '.yaml', '.json', '.md', '.markdown', 'image/gif', 'image/jpg', 'image/jpeg', 'image/ico', 'image/png']" appearance="file" removable />
```

### help

Use the `help` prop to display text under KInput.

<KCard>
  <template v-slot:body>

>     <KFileUpload label="Upload File" :label-attributes="{ help: 'Hello, here is the example' }" help="Some text here.." icon="close" :accept="['.yml', '.yaml', '.json', '.md', '.markdown', 'image/gif', 'image/jpg', 'image/jpeg', 'image/ico', 'image/png']" appearance="file" removable />

Reviewer: ```suggestion
    <KFileUpload label="Upload File" help="Some text here.." :accept="['.yml', '.yaml', '.json', '.md', '.markdown']" appearance="file" removable />
```

Can we simplify?

  </template>
</KCard>

```html
<KFileUpload label="Upload File" help="Some text here.." icon="close" :accept="['.yml', '.yaml', '.json', '.md', '.markdown', 'image/gif', 'image/jpg', 'image/jpeg', 'image/ico', 'image/png']" appearance="file" removable />
```

### placeholder

> Use the `placeholder` prop to display placeholder text.

Reviewer: May want to call out that the placeholder text is blue to indicate the field is clickable? May be overkill, but we are kind of treating this differently than most usages of `placeholder`

Reviewer: Also, I think the placeholder text should be overwritten with the file name once one has been accepted

<KCard class="mt-6">
  <template v-slot:body>

>     <KFileUpload appearance="image" :accept="['.yml', '.yaml', '.json', '.md', '.markdown', 'image/gif', 'image/jpg', 'image/jpeg', 'image/ico', 'image/png']" placeholder="You can change the text here!" icon="kong" >

Reviewer: ```suggestion
    <KFileUpload appearance="image" :accept="['image/gif', 'image/jpg', 'image/jpeg', 'image/ico', 'image/png']" placeholder="You can change the text here!" icon="kong" >
```

Can we simplify?

    </KFileUpload>
  </template>
</KCard>

```html
<KFileUpload appearance="image" :accept="['.yml', '.yaml', '.json', '.md', '.markdown', 'image/gif', 'image/jpg', 'image/jpeg', 'image/ico', 'image/png']" placeholder="You can change the text here!" icon="kong" />
```

### removable

> A `cancel` button can be displayed for `file` type upload, by default this is set to `false`. This button is only visible once a file has been added.

Reviewer: Again, I wouldn't restrict this prop to being used with `file` appearance. Users can choose whether or not they want to use it. Can we get an example?

Reviewer: Also think it may be good to have this `true` by default as I think most cases are going to want the option to remove the file once it's been selected. What do you think?

### accept

Please refer to [mdn web docs](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/file#accept) that defines the file types the file input should accept. 

### buttonAppearance

> Use this prop to customize the trigger **KButton** [appearance](/components/button.html#appearance). Default set to `primary`.

Reviewer: May want to call out that this button only shows for `file` appearance?

<KCard>
  <template v-slot:body>

>     <KFileUpload label="Upload File" buttonAppearance="creation" :label-attributes="{ help: 'I use the KLabel `help` prop' }" help="Additional text." icon="close" :accept="['.yml', '.yaml', '.json', '.md', '.markdown', 'image/gif', 'image/jpg', 'image/jpeg', 'image/ico', 'image/png']" appearance="file" removable />

Reviewer: ```suggestion
    <KFileUpload label="Upload File" buttonAppearance="creation" :accept="['.yml', '.yaml', '.json', '.md', '.markdown']" appearance="file" removable />
```

Can we simplify?

  </template>
</KCard>

```html
<KFileUpload label="Upload File" buttonAppearance="creation" :label-attributes="{ help: 'I use the KLabel `help` prop' }" help="Additional text." icon="close" :accept="['.yml', '.yaml', '.json', '.md', '.markdown', 'image/gif', 'image/jpg', 'image/jpeg', 'image/ico', 'image/png']" appearance="file" removable />
```

### buttonText

> This is the text that will be displayed on the button that triggers the click on KInput.

Reviewer: Same comments as `buttonAppearance`

<KCard>
  <template v-slot:body>
    <KFileUpload label="Upload File" buttonAppearance="danger" buttonText="Click me" :label-attributes="{ help: 'I use the KLabel `help` prop' }" help="Additional text." icon="close" :accept="['.yml', '.yaml', '.json', '.md', '.markdown', 'image/gif', 'image/jpg', 'image/jpeg', 'image/ico', 'image/png']" appearance="file" removable />
  </template>
</KCard>

```html
<KFileUpload label="Upload File" buttonAppearance="danger" buttonText="Click me" :label-attributes="{ help: 'I use the KLabel `help` prop' }" help="Additional text." icon="close" :file-input-accept-types="['.yml', '.yaml', '.json', '.md', '.markdown', 'image/gif', 'image/jpg', 'image/jpeg', 'image/ico', 'image/png']" appearance="file" removable />
```

### icon

Specify an icon to display to the left of the `placeholder text` if `appearance='image'`.

### iconSize

The size of the `icon` being displayed (default is `26`).

### iconColor

> The color of the `icon` being displayed.

Reviewer: Can we show one example using all 3 icon related props?

### maxFileSize

> Use this prop to customize the maximize size of file that can be uploaded. Default value for `image` upload is `1mb (1000000 bytes)` and for `any other file` is `5.24mb (5242880 bytes)`.

Reviewer: ```suggestion
Use this prop to customize the maximize size of file that can be uploaded. Default value for `image` upload is `1mb (1000000 bytes)` and for `file` is `5.24mb (5242880 bytes)`.
```

Can we tell them what happens if they try to upload a file larger than that?

Author: it emits 'file-upload-error', do you want me to add an 'example' in 'md' file ?

Reviewer 2: @aanchalm01 instead of `file-upload-error` can we just emit `error` that would be used like this? `<KFileUpload @error="errorCallback" />`

## Slots

- `icon` - Icon to the left of the placeholder text inside KInput for `appearance ='image'`

## Events

The events below will fire whenever:
- a file is added
- a file is removed
- there is a file upload error

<KCard>
  <template v-slot:body>

>     <KFileUpload label="Upload File" appearance="file" :label-attributes="{ help: 'I use the KLabel `help` prop' }" help="Additional files can be uploaded from HomePage." icon="close" :accept="['.yml', '.yaml', '.json', '.md', '.markdown', 'image/gif', 'image/jpg', 'image/jpeg', 'image/ico', 'image/png']" removable @file-added="i => printData(i)" @file-removed="() => { fileData = '' }" />

Author: added 'image/*' instead of all the file types

/>
```

<KCard class="mt-6">
  <template v-slot:body>
    <KFileUpload type="image" label="Upload Image File" :label-attributes="{ help: `Accepted file types: ${acceptedImageType}` }" class="image-with-label" icon="image" :accept="acceptedImageType" placeholder="Upload new image (Max 1 MB)" />
  </template>
</KCard>

```html
<KFileUpload 
  type="image" 
  label="Upload Image file" 
  :label-attributes="{ help: 'Accepted file types: image/*' }"
  icon="image" 
  placeholder="Upload new image (Max 1 MB)"
/>
```

### help

Use the `help` prop to display text under KInput.

<KCard>
  <template v-slot:body>
    <KFileUpload label="Upload File" :label-attributes="{ help: `Accepted file types: ${acceptedFileType}` }" help="Some text here.." :accept="acceptedFileType" />
  </template>
</KCard>

```html
<KFileUpload label="Upload File" help="Some text here.." />
```

### hasError

Boolean value to indicate whether the element has an error and should apply error styling. By default this is `false`.

### errorMessage

String to be displayed as error message if `hasError` prop is `true`.

```html
<KFileUpload
  hasError
  errorMessage="File size should be less than 1 mb."
/>

> ```

Reviewer 2: Can we show a working example of the `errorMessage` ?

### placeholder

Use the `placeholder` prop to display placeholder text. The `placeholder text` is `blue` to indicate the field is `clickable`.

<KCard class="mt-6">
  <template v-slot:body>
    <KFileUpload type="image" label="Upload File" :label-attributes="{ help: `Accepted file types: ${acceptedImageType}` }" class="image-with-label" :accept="acceptedImageType" placeholder="You can change the text here!" icon="kong" >
    </KFileUpload>
  </template>
</KCard>

```html
<KFileUpload 
  type="image" 
  label="Upload File"
  placeholder="You can change the text here!" 
  icon="kong" 
/>
```

### removable

> A `cancel` button can be displayed, by default this is set to `true`. This button is only visible once a file has been added.

Reviewer 2: question: Why would we not want to let a user remove a file?

<KCard>
  <template v-slot:body>
    <KFileUpload label="Upload File" :label-attributes="{ help: `Accepted file types: ${acceptedFileType}` }" buttonAppearance="danger" buttonText="Click me" :accept="acceptedFileType" />
  </template>
</KCard>

```html
<KFileUpload label="Upload File" buttonAppearance="danger" buttonText="Click me" />
```

### icon

Specify an icon to display to the left of the `placeholder text` if prop `type` is `image`.

### iconSize

The size of the `icon` being displayed (default is `26`).

### iconColor

The color of the `icon` being displayed.

<KCard class="mt-6">
  <template v-slot:body>
    <KFileUpload type="image" label="Upload File" :label-attributes="{ help: `Accepted file types: ${acceptedImageType}` }" :accept="acceptedImageType" class="image-with-label" placeholder="Customized icon, iconColor & iconSize!" icon="immunity" iconColor="gold" iconSize="30" />
  </template>
</KCard>

```html
<KFileUpload 
  type="image" icon="immunity" iconColor="gold" iconSize="30" 
  placeholder="You can change the text here!"
/>
```
### maxFileSize

> Use this prop to customize the maximize size of file that can be uploaded. Default value for `image` upload is `1mb (1000000 bytes)` and for `file` is `5.24mb (5242880 bytes)`.

Reviewer 2: nit: I don't think we need to call out the size conversion, and the abbreviation should be uppercase.

Also, why `5.24MB` for files? That seems very specific; why not just `5MB`?

```suggestion
Use this prop to customize the maximize size of file that can be uploaded. Default value for `image` upload is `1MB` and for `file` is `5.24MB`.
```

Author: `khcp-ui` `DocUploadCard` uses 5.24MB as max file size; that's why I selected that as the max size.

## Events

The events below will fire whenever:

- a file is added `file-added`
- a file is removed `file-removed`
- there is a file upload error `error`

<KCard>
  <template v-slot:body>
    <KFileUpload label="Upload File" :label-attributes="{ help: `Accepted file types: ${acceptedFileType}` }" :accept="acceptedFileType" @file-added="file => printData(file)" @file-removed="() => { fileData = '' }" />
  </template>

> </KCard>

Reviewer 2: You need to put the example code below this block so people can see how it's being used in the docs.
